smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/documentation/StructureExampleGenerator.java

```diff
@@ -261,23 +263,41 @@ private static void append(int indentation, StringBuilder buffer, String tail) {
      * This handles the case of recursive shapes.
      */
     private static class ShapeTracker {
-        private Map<Shape, Set<Integer>> data = new HashMap<Shape, Set<Integer>>();
+        private Map<Shape, Set<Integer>> depths = new HashMap<Shape, Set<Integer>>();
+        private Map<Shape, Integer> occurrences = new HashMap<Shape, Integer>();
 
         /**
          * Mark that a shape is observed at depth.
          */
         public void mark(Shape shape, int depth) {
-            if (!data.containsKey(shape)) {
-                data.put(shape, new HashSet<>());
+            if (!depths.containsKey(shape)) {
+                depths.put(shape, new HashSet<>());
             }
-            data.get(shape).add(depth);
+            depths.get(shape).add(depth);
+            occurrences.put(shape, occurrences.getOrDefault(shape, 0) + 1);
+        }
+
+        /**
+         * @return whether the shape should be truncated.
+         */
+        public boolean shouldTruncate(Shape shape) {
+            return (shape instanceof MapShape || shape instanceof UnionShape || shape instanceof StructureShape
+                    || shape instanceof ListShape || shape instanceof SetShape)
+                    && (getOccurrenceCount(shape) > 2 || getOccurrenceDepths(shape) > 2);
```

Comment:
technically occurrence count is greater or equal to depth count, but I left it there in case it needs further adjustments.